### PR TITLE
Fix bugs in agentendpoint code

### PR DESCRIPTION
**Pull request by adjackura**
Add more debug logging
Add LookupEffectiveGuestPolicies call, currently unused
Sleep the poll interval between calls (10m) if the osconfig service in not enabled
Sleep longer between API call retries on ResourceExhausted errors
Fix race in tests
Decode jwt token correctly

### CODE REVIEW

File: agentendpoint/agentendpoint.go [hunk 1]
```diff
@@ -43,26 +49,59 @@ type Client struct {
...
+
+// Close closes the underlying ClientConn.
+func (c *Client) Close() error {
+	return c.raw.Close()
 }
 
-// ReportTaskProgress calls the agentendpoint service ReportTaskProgress call injecting the InstanceIdToken.
-func (c *Client) ReportTaskProgress(ctx context.Context, req *agentendpointpb.ReportTaskProgressRequest) (res *agentendpointpb.ReportTaskProgressResponse, err error) {
-	logger.Debugf("ReportTaskProgress request:\n%s", util.PrettyFmt(req))
+func (c *Client) reportTaskStart(ctx context.Context) (res *agentendpointpb.ReportTaskStartResponse, err error) {
```
Comment (jaiminsh) [inline agentendpoint/agentendpoint.go]: s/reportTaskStart/ReportTaskStart

File: agentendpoint/agentendpoint.go [hunk 2]
```diff
@@ -43,26 +49,59 @@ type Client struct {
...
+		logger.Debugf("ReportTaskStart response:\n%s", util.PrettyFmt(res))
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("error calling ReportTaskStart: %v", err)
+	}
+
+	return
+}
+
+func (c *Client) reportTaskProgress(ctx context.Context, req *agentendpointpb.ReportTaskProgressRequest) (res *agentendpointpb.ReportTaskProgressResponse, err error) {
```
Comment (jaiminsh) [inline agentendpoint/agentendpoint.go]: s/reportTaskProgress/ReportTaskProgress

File: agentendpoint/agentendpoint.go [hunk 3]
```diff
@@ -43,26 +49,59 @@ type Client struct {
...
 
-var client *Client
-
 // NewClient a new agentendpoint Client.
 func NewClient(ctx context.Context) (*Client, error) {
-	c, err := agentendpoint.NewClient(ctx)
+	opts := []option.ClientOption{
+		option.WithoutAuthentication(),
+		option.WithGRPCDialOption(grpc.WithTransportCredentials(credentials.NewTLS(nil))),
+		option.WithEndpoint(config.SvcEndpoint()),
```
Comment (jaiminsh) [inline agentendpoint/agentendpoint.go]: Can you add a comment describing why these options are required?

File: agentendpoint/agentendpoint.go [hunk 4]
```diff
@@ -99,36 +139,65 @@ func (c *Client) ReportTaskComplete(ctx context.Context, req *agentendpointpb.Re
...
+		OsVersion:      info.Version,
+		OsArchitecture: info.Architecture,
+	}
+
+	token, err := config.IDToken()
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Debugf("LookupEffectiveGuestPolicies request:\n%s", util.PrettyFmt(req))
```
Comment (jaiminsh) [inline agentendpoint/agentendpoint.go]: Calling LookupEffectiveGuestPolicies with request: to be consistent with other logging statements.

File: agentendpoint/agentendpoint.go (hunk 1 above)
Comment (adjackura) [reply agentendpoint/agentendpoint.go]: I made these private as there's no need to access them outside the package

File: agentendpoint/agentendpoint.go [hunk 5]
```diff
@@ -199,19 +271,27 @@ func (c *Client) loadTaskFromState(ctx context.Context) error {
...
 
 		if err := c.watchStream(ctx, noti, token); err != nil {
-			return err
+			if s, ok := status.FromError(err); ok && s.Code() == codes.PermissionDenied {
+				// Service is not enabled for this project.
+				time.Sleep(config.SvcPollInterval())
+				continue
+			}
+			logger.Errorf("Error watching stream: %v", err)
+			time.Sleep(5 * time.Second)
```
Comment (ryanwe) [inline agentendpoint/agentendpoint.go]: Can you add a TODO to add backoff? This could be problematic without it.

Comment (adjackura) [reply agentendpoint/agentendpoint.go]: Done, this can get complicated, we may want to add different waits for many different classes of errors.

**Pull request merged**